TulparAudio/source/TulparConfigurator.cpp

```diff
@@ -0,0 +1,73 @@
+/*
+* Copyright (C) 2018 by Godlike
+* This code is licensed under the MIT license (MIT)
+* (http://opensource.org/licenses/MIT)
+*/
+
+#include <tulpar/TulparConfigurator.hpp>
+
+#include <tulpar/InternalLoggers.hpp>
+
+#include <AL/alc.h>
+
+namespace tulpar
+{
+
+TulparConfigurator::TulparConfigurator()
+    : bufferBatch(32)
+    , sourceBatch(32)
+    , device(TulparConfigurator::Device
+        {
+            std::string()
+            , true
+        })
+{
+
+}
+
+std::vector<TulparConfigurator::Device> TulparConfigurator::GetDevices()
+{
+    std::vector<TulparConfigurator::Device> result;
+
+    std::string const defaultDevice(alcGetString(NULL, ALC_DEFAULT_ALL_DEVICES_SPECIFIER));
+    ALCchar const* devices = alcGetString(NULL, ALC_ALL_DEVICES_SPECIFIER);
+
+    if (NULL != devices && '\0' != (*devices))
+    {
+        ALCchar const* it = devices;
+
+        std::string name(it);
+        it += name.size() + 1;
+
+        result.push_back({name, (0 == defaultDevice.compare(name))});
+
+        while (*it != '\0')
+        {
+            name = it;
+            it += name.size() + 1;
+
+            result.push_back({name, (0 == defaultDevice.compare(name))});
+        }
+
+        LOG->Debug("TulparConfigurator::GetDevices() found {} devices", result.size());
```

Comment:
Is there a way to disable logging? In case of shipping might be useful.

Reply:
Not really. Mule could be extended to evaluate to some `void` expressions instead of an actual logging, but the signature would have to stay.
`spdlog` that is used underneath has some macros to enable/disable certain levels of logging during compilation, it can be mimicked in Mule if needed.

I do not believe that compile-level debug configuration would be that valuable/used. For runtime there is possibility to control log verbosity and you can disable logging there. It would still cause some jumps and compares but it shouldn't be a big deal.